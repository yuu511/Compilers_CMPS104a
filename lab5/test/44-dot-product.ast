ROOT "<<ROOT>>" 0.0.0 {0} 
FUNCTION "" 6.22.1 {0} void function 
TYPE_ID "" 6.22.1 {0} 
VOID "void" 6.22.1 {0} 
IDENT "fail" 6.22.6 {0} 
PARAM "(" 6.22.11 {0} 
TYPE_ID "" 6.22.12 {1} string variable param lval 
STRING "string" 6.22.12 {0} 
IDENT "expr" 6.22.19 {0} 
TYPE_ID "" 6.22.25 {1} string variable param lval 
STRING "string" 6.22.25 {0} 
IDENT "file" 6.22.32 {0} 
TYPE_ID "" 6.22.38 {1} int variable param lval 
INT "int" 6.22.38 {0} 
IDENT "line" 6.22.42 {0} 
FUNCTION "" 6.24.1 {1} void function 
TYPE_ID "" 6.24.1 {0} 
VOID "void" 6.24.1 {0} 
IDENT "putchr" 6.24.6 {0} 
PARAM "(" 6.24.13 {0} 
TYPE_ID "" 6.24.14 {2} int variable param lval 
INT "int" 6.24.14 {0} 
IDENT "chr" 6.24.18 {0} 
FUNCTION "" 6.25.1 {2} void function 
TYPE_ID "" 6.25.1 {0} 
VOID "void" 6.25.1 {0} 
IDENT "putint" 6.25.6 {0} 
PARAM "(" 6.25.13 {0} 
TYPE_ID "" 6.25.14 {3} int variable param lval 
INT "int" 6.25.14 {0} 
IDENT "num" 6.25.18 {0} 
FUNCTION "" 6.26.1 {3} void function 
TYPE_ID "" 6.26.1 {0} 
VOID "void" 6.26.1 {0} 
IDENT "putstr" 6.26.6 {0} 
PARAM "(" 6.26.13 {0} 
TYPE_ID "" 6.26.14 {4} string variable param lval 
STRING "string" 6.26.14 {0} 
IDENT "str" 6.26.21 {0} 
FUNCTION "" 6.28.1 {4} int function 
TYPE_ID "" 6.28.1 {0} 
INT "int" 6.28.1 {0} 
IDENT "getchr" 6.28.5 {0} 
PARAM "(" 6.28.11 {0} 
FUNCTION "" 6.29.1 {5} string function 
TYPE_ID "" 6.29.1 {0} 
STRING "string" 6.29.1 {0} 
IDENT "getstr" 6.29.8 {0} 
PARAM "(" 6.29.14 {0} 
FUNCTION "" 6.30.1 {6} string function 
TYPE_ID "" 6.30.1 {0} 
STRING "string" 6.30.1 {0} 
IDENT "getln" 6.30.8 {0} 
PARAM "(" 6.30.13 {0} 
FUNCTION "" 7.5.1 {7} int function 
TYPE_ID "" 7.5.1 {0} 
INT "int" 7.5.1 {0} 
IDENT "dot_product" 7.5.5 {0} 
PARAM "(" 7.5.17 {0} 
TYPE_ID "" 7.5.18 {8} int variable param lval 
INT "int" 7.5.18 {0} 
IDENT "size" 7.5.22 {0} 
TYPE_ID "" 7.5.28 {8} int array variable param lval 
ARRAY "array" 7.5.28 {0} 
INT "int" 7.5.34 {0} 
IDENT "vec1" 7.5.39 {0} 
TYPE_ID "" 7.5.45 {8} int array variable param lval 
ARRAY "array" 7.5.45 {0} 
INT "int" 7.5.51 {0} 
IDENT "vec2" 7.5.56 {0} 
BLOCK "{" 7.5.62 {8} 
TYPE_ID "" 7.6.4 {8} int variable local lval vreg (7.6.4)
INT "int" 7.6.4 {0} 
IDENT "index" 7.6.8 {0} 
INTCON "0" 7.6.16 {8} int const 
TYPE_ID "" 7.7.4 {8} int variable local lval vreg (7.7.4)
INT "int" 7.7.4 {0} 
IDENT "dot" 7.7.8 {0} 
INTCON "0" 7.7.14 {8} int const 
WHILE "while" 7.8.4 {8} 
LT "<" 7.8.17 {8} int vreg 
IDENT "index" 7.8.11 {8} int variable local lval (7.6.4)
IDENT "size" 7.8.19 {8} int variable param lval (7.5.18)
BLOCK "{" 7.8.25 {8} 
'=' "=" 7.9.11 {8} int variable local lval vreg 
IDENT "dot" 7.9.7 {8} int variable local lval (7.7.4)
'+' "+" 7.9.17 {8} int vreg 
IDENT "dot" 7.9.13 {8} int variable local lval (7.7.4)
'*' "*" 7.9.31 {8} int vreg 
INDEX "[" 7.9.23 {8} int lval vaddr 
IDENT "vec1" 7.9.19 {8} int array variable param lval (7.5.28)
IDENT "index" 7.9.24 {8} int variable local lval (7.6.4)
INDEX "[" 7.9.37 {8} int lval vaddr 
IDENT "vec2" 7.9.33 {8} int array variable param lval (7.5.45)
IDENT "index" 7.9.38 {8} int variable local lval (7.6.4)
'=' "=" 7.10.13 {8} int variable local lval vreg 
IDENT "index" 7.10.7 {8} int variable local lval (7.6.4)
'+' "+" 7.10.21 {8} int vreg 
IDENT "index" 7.10.15 {8} int variable local lval (7.6.4)
INTCON "1" 7.10.23 {8} int const 
RETURN "return" 7.12.4 {8} 
IDENT "dot" 7.12.11 {8} int variable local lval (7.7.4)
FUNCTION "" 7.17.1 {8} int function 
TYPE_ID "" 7.17.1 {0} 
INT "int" 7.17.1 {0} 
IDENT "main" 7.17.5 {0} 
PARAM "(" 7.17.9 {0} 
BLOCK "{" 7.17.12 {9} 
TYPE_ID "" 7.18.4 {9} int array variable local lval vreg (7.18.4)
ARRAY "array" 7.18.4 {0} 
INT "int" 7.18.10 {0} 
IDENT "vec1" 7.18.15 {0} 
ALLOC "alloc" 7.18.22 {9} int array 
ARRAY "array" 7.18.28 {0} 
INT "int" 7.18.34 {0} 
INTCON "10" 7.18.41 {9} int const 
TYPE_ID "" 7.19.4 {9} int array variable local lval vreg (7.19.4)
ARRAY "array" 7.19.4 {0} 
INT "int" 7.19.10 {0} 
IDENT "vec2" 7.19.15 {0} 
ALLOC "alloc" 7.19.22 {9} int array 
ARRAY "array" 7.19.28 {0} 
INT "int" 7.19.34 {0} 
INTCON "10" 7.19.41 {9} int const 
TYPE_ID "" 7.20.4 {9} int variable local lval vreg (7.20.4)
INT "int" 7.20.4 {0} 
IDENT "index" 7.20.8 {0} 
INTCON "0" 7.20.16 {9} int const 
WHILE "while" 7.21.4 {9} 
LT "<" 7.21.17 {9} int vreg 
IDENT "index" 7.21.11 {9} int variable local lval (7.20.4)
INTCON "10" 7.21.19 {9} int const 
BLOCK "{" 7.21.23 {9} 
'=' "=" 7.22.19 {9} int lval vreg vaddr 
INDEX "[" 7.22.11 {9} int lval vaddr 
IDENT "vec1" 7.22.7 {9} int array variable local lval (7.18.4)
IDENT "index" 7.22.12 {9} int variable local lval (7.20.4)
'+' "+" 7.22.27 {9} int vreg 
IDENT "index" 7.22.21 {9} int variable local lval (7.20.4)
INTCON "10" 7.22.29 {9} int const 
'=' "=" 7.23.19 {9} int lval vreg vaddr 
INDEX "[" 7.23.11 {9} int lval vaddr 
IDENT "vec2" 7.23.7 {9} int array variable local lval (7.19.4)
IDENT "index" 7.23.12 {9} int variable local lval (7.20.4)
'*' "*" 7.23.27 {9} int vreg 
IDENT "index" 7.23.21 {9} int variable local lval (7.20.4)
INTCON "10" 7.23.29 {9} int const 
'=' "=" 7.24.13 {9} int variable local lval vreg 
IDENT "index" 7.24.7 {9} int variable local lval (7.20.4)
'+' "+" 7.24.21 {9} int vreg 
IDENT "index" 7.24.15 {9} int variable local lval (7.20.4)
INTCON "1" 7.24.23 {9} int const 
CALL "(" 7.26.11 {2} void function (6.25.1)
IDENT "putint" 7.26.4 {0} 
CALL "(" 7.26.24 {7} int function (7.5.1)
IDENT "dot_product" 7.26.12 {0} 
INTCON "10" 7.26.25 {9} int const 
IDENT "vec1" 7.26.29 {9} int array variable local lval (7.18.4)
IDENT "vec2" 7.26.35 {9} int array variable local lval (7.19.4)
CALL "(" 7.27.11 {1} void function (6.24.1)
IDENT "putchr" 7.27.4 {0} 
CHARCON "'\n'" 7.27.12 {9} int const 
RETURN "return" 7.28.4 {9} 
INTCON "0" 7.28.11 {9} int const 
