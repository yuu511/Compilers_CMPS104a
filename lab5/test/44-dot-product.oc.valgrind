==5928== Memcheck, a memory error detector
==5928== Copyright (C) 2002-2017, and GNU GPL'd, by Julian Seward et al.
==5928== Using Valgrind-3.13.0 and LibVEX; rerun with -h for copyright info
==5928== Command: ../oc 44-dot-product.oc
==5928== 
==5928== 
==5928== HEAP SUMMARY:
==5928==     in use at exit: 0 bytes in 0 blocks
==5928==   total heap usage: 865 allocs, 865 frees, 169,896 bytes allocated
==5928== 
==5928== All heap blocks were freed -- no leaks are possible
==5928== 
==5928== For counts of detected and suppressed errors, rerun with: -v
==5928== ERROR SUMMARY: 0 errors from 0 contexts (suppressed: 0 from 0)
