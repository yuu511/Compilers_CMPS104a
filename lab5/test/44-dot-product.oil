dot_product: .function int 
           .param int size
           .param ptr vec1
           .param ptr vec2
           .local int index
           .local int dot
           index = 0  
           dot = 0  
.wh0:      goto .od0 if index > size
.do0:      $t2:i = vec1[index * :i]  
           $t3:i = vec2[index * :i]  
           $t1:i = $t2:i * $t3:i
           $t0:i = dot + $t1:i
           dot = $t0:i  
           $t4:i = index + 1
           index = $t4:i  
           goto .wh0  
.od0
           return dot
           return
           .end
main:      .function int 
           .local ptr vec1
           .local ptr vec2
           .local int index
           $t0:p = 10 * sizeof int
           vec1 = call malloc($t0:p)  
           $t1:p = 10 * sizeof int
           vec2 = call malloc($t1:p)  
           index = 0  
.wh1:      goto .od1 if index > 10
.do1:      $t2:i = index + 10
           vec1[index * :i] = $t2:i  
           $t3:i = index * 10
           vec2[index * :i] = $t3:i  
           $t4:i = index + 1
           index = $t4:i  
           goto .wh1  
.od1
           $t5:i = call dot_product(10,vec1,vec2)  
           call putint($t5:i)  
           call putchr('\n')  
           return 0
           return
           .end
