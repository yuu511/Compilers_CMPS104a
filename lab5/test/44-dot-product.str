string_set[   0]:    7296690267450279874 0x55d086419d58->"chr"
                   13348923631108115385 0x55d08641a718->"putstr"
string_set[   5]:   12679825168387489523 0x55d08641a278->"num"
string_set[   7]:    6814028903552918164 0x55d08641b8b8->"array"
                    2297668033614959926 0x55d08641c158->"0"
string_set[  10]:    6637313742931709005 0x55d08641cf78->"*"
                    1583576014053361350 0x55d08641a8d8->"str"
                   13497578857207247412 0x55d086419318->"string"
                   11597697714117577063 0x55d08641c098->"="
string_set[  11]:    1396094175042499165 0x55d08641b978->"<"
string_set[  13]:    2594088158685378223 0x55d08641e018->"10"
                   12868445110721718657 0x55d086419978->")"
string_set[  16]:   10159970873491820195 0x55d08641d438->"1"
                   11911796782241914524 0x55d086418088->"void"
                    4698864730501174006 0x55d08641bfd8->"index"
string_set[  17]:   15298581498043935097 0x55d08641c2c8->"dot"
string_set[  18]:   13581035385482130682 0x55d08641cb58->"["
string_set[  19]:    5344662657899890615 0x55d086419a38->";"
string_set[  20]:   11868120863286193964 0x55d08641d798->"main"
                   15816506932648033945 0x55d086419258->"("
string_set[  24]:   13934964484049229816 0x55d08641b6d8->"size"
string_set[  25]:   11166701219533395393 0x55d08641aef8->"getstr"
string_set[  29]:   14635384904416252381 0x55d08641d5e8->"return"
                    1370730858159036685 0x55d08641ca18->"+"
string_set[  34]:    8290819294854170237 0x55d08641dbd8->"alloc"
                    7211130878751977441 0x55d086419b98->"putchr"
string_set[  37]:    6192799629375248351 0x55d08641f848->"'\n'"
                   12291392672316406109 0x55d086419898->"line"
string_set[  39]:   18413295556186398326 0x55d08641a0b8->"putint"
string_set[  40]:   10319632053171904911 0x55d08641c558->"while"
string_set[  43]:   11374676767421005072 0x55d08641b518->"dot_product"
string_set[  44]:      21606623943142546 0x55d08641be98->"{"
string_set[  45]:    6142509188972423790 0x55d086419498->""
string_set[  47]:   13686721134187162920 0x55d08641d558->"}"
string_set[  49]:    9330387579099533542 0x55d08641bd98->"vec2"
string_set[  53]:   12051031859636400188 0x55d08641cf18->"]"
string_set[  57]:    9386170080689500517 0x55d0864193d8->"expr"
string_set[  58]:    4017852018407011111 0x55d086419658->"file"
                    3729804957429652673 0x55d086412d08->"<<ROOT>>"
string_set[  59]:   16579916199511856337 0x55d08641b1f8->"getln"
string_set[  63]:    9611428142580219387 0x55d086419158->"fail"
string_set[  64]:   12597418120160561823 0x55d0864197d8->"int"
string_set[  67]:    2486443799175530694 0x55d08641bb18->"vec1"
string_set[  70]:    9201800968376702500 0x55d08641abf8->"getchr"
string_set[  71]:   14526079891305879776 0x55d08641bab8->">"
string_set[  78]:    4706206694200543916 0x55d086419198->","
load_factor = 0.582
bucket_count = 79
max_bucket_size = 4
