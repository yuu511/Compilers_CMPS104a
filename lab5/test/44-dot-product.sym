main (7.17.1) {8} int function 
   vec1 (7.18.4) {9} int array variable local lval 0
   vec2 (7.19.4) {9} int array variable local lval 1
   index (7.20.4) {9} int variable local lval 2

getln (6.30.1) {6} string function 

putchr (6.24.1) {1} void function 

dot_product (7.5.1) {7} int function 
   size (7.5.18) {8} int variable param lval 0
   vec1 (7.5.28) {8} int array variable param lval 1
   vec2 (7.5.45) {8} int array variable param lval 2
   index (7.6.4) {8} int variable local lval 3
   dot (7.7.4) {8} int variable local lval 4

getstr (6.29.1) {5} string function 

fail (6.22.1) {0} void function 

putint (6.25.1) {2} void function 

putstr (6.26.1) {3} void function 

getchr (6.28.1) {4} int function 

