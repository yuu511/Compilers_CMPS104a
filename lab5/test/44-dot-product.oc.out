CONDITIONAL WHILE STATEMENT SYMBOL TOK_LT 
CONDITIONAL WHILE STATEMENT SYMBOL TOK_LT 
